File: semaforos/exemploLivro/rendezvous/Rendezvous.java
package semaforos.exemploLivro.rendezvous;

import java.util.concurrent.Semaphore;

public class Rendezvous implements Runnable{

    private static Semaphore a1Arrive = new Semaphore(0);
    private static Semaphore b1Arrive = new Semaphore(0);

    public static void main(String[] args) {
        Thread a1 = new Thread(new Rendezvous(),"A1");
        Thread b1 = new Thread(new Rendezvous(),"B1");
        Thread a2 = new Thread(new Rendezvous(),"A2");
        Thread b2 = new Thread(new Rendezvous(),"B2");

        a1.start();
        a2.start();
        b1.start();
        b2.start();

        try {
            a1.join();
            a2.join();
            b1.join();
            b2.join();
        } catch (InterruptedException e) {
            System.err.println(e.getStackTrace());
        }
    }

    @Override
    public void run() {
        String name = Thread.currentThread().getName();

        switch (name) {
            case "A1":
                System.out.println("A1 Chegou!");
                a1Arrive.release();
                break;
            case "B1":
                System.out.println("B1 Chegou!");
                b1Arrive.release();
                break;
            case "A2":
                try {
                    b1Arrive.acquire();
                } catch (InterruptedException e) {
                    System.err.println(e.getStackTrace());
                }
                System.out.println("A2 Chegou!");
                break;
            case "B2":
                try {
                    a1Arrive.acquire();
                } catch (InterruptedException e) {
                    System.err.println(e.getStackTrace());
                }
                System.out.println("B2 Chegou!");
                break;
            default:
                break;
        }
    }
    
}

File: semaforos/exemploLivro/rendezvous/rendezvous.py
import threading

aArrive = threading.Semaphore(0)
bArrive = threading.Semaphore(0)

def exibeA1():
    print("A1 Chegou")
    aArrive.release()

def exibeB1():
    print("B1 Chegou")
    bArrive.release()

def exibeB2():
    aArrive.acquire()
    print("B2 Chegou")

def exibeA2():
    bArrive.acquire()
    print("A2 Chegou")

if __name__ == "__main__":

    a1 = threading.Thread(target=exibeA1,name="A1")
    b1 = threading.Thread(target=exibeB1,name="B1")
    a2 = threading.Thread(target=exibeB2,name="B2")
    b2 = threading.Thread(target=exibeA2,name="A2")

    a1.start()
    a2.start()
    b1.start()
    b2.start()

    a1.join()
    a2.join()
    b1.join()
    b2.join()

